### ISSUE TITLE
[io.fits] Position of the EXTEND keyword

### ISSUE THREAD

**Issue by olebole** [astropy/astropy#5883]
Hi,

while packaging parts of the [Starjava](https://github.com/Starlink/starjava) software to get TOPCAT and STILTS into Debian, I stumbled upon their so-called [FITS-Plus](http://www.star.bristol.ac.uk/~mbt/topcat/sun253/outFits.html) and [colfits-plus](http://www.star.bristol.ac.uk/~mbt/topcat/sun253/outColfits.html) formats. These formats are basically [recognized](http://www.star.bristol.ac.uk/~mbt/stil/javadocs/uk/ac/starlink/votable/FitsPlusTableWriter.html) by the keywords `COLFITS` (for colfits-plus) and `VOTMETA` keywords _before the `EXTEND` keyword_ (order [matters](https://github.com/Starlink/starjava/blob/e27e8f0f86d313cde9ac110af62b0a43a15bf873/votable/src/main/uk/ac/starlink/votable/FitsPlusTableWriter.java#L68-L69) here!):
```
SIMPLE  =                    T / Standard FITS format
BITPIX  =                    8 / Character data
NAXIS   =                    1 / Text string
NAXIS1  =                 nnnn / Number of characters
VOTMETA =                    T / Table metadata in VOTABLE format
EXTEND  =                    T / There are standard extensions
```
While the primary problem here is to get this supported by the Java IO package `nom.tam.fits` (see nom-tam-fits/nom-tam-fits#113), I am wondering about the support in Astropy. Currently, the header are re-sorted so that `EXTEND` directly follows `NAXIS*` (similarly to currently `nom.tam.fits`):
```
>>> from astropy.io import fits
>>> p = fits.PrimaryHDU()
>>> p.header.set('VOTMETA', True, before='EXTEND')
>>> print(list(p.header))
['SIMPLE', 'BITPIX', 'NAXIS', 'VOTMETA', 'EXTEND']
>>> p.writeto('sample.fits')
>>> print(list(p.header))
['SIMPLE', 'BITPIX', 'NAXIS', 'EXTEND', 'VOTMETA']
```
In the current FITS definition, I couldnt find the requirement that `EXTEND` should directly follow `NAXIS*`; however it would probably be useful if the different FITS libraries are somehow consistent.
Would it be possible to support these formats by Astropy? And what would be the optimal way to do this?

**Comment by MSeifert04** [comment]
At least according to http://heasarc.gsfc.nasa.gov/docs/fcg/standard_dict.html:

> DEFINITION:  If the FITS file may contain extensions, a card image with
the keyword EXTEND and the value field containing the logical value T
**must appear in the  primary header immediately after the last NAXISn
card image, or, if NAXIS=0, the NAXIS card image**.  The presence of this
keyword with the value T in the primary header does not require that
extensions be present.

That adresses only part of your question. But I thought I mention it.

**Comment by olebole** [comment]
Yea, that was 1999, and this requirement was there also in [2.1b (2005)](https://fits.gsfc.nasa.gov/standard21b/fits_standard21b.pdf). However, in [3.0](https://fits.gsfc.nasa.gov/standard30/fits_standard30aa.pdf), this requirement was removed (together with the move of `EXTEND` from the list of required keywords to the list of reserved keywords).

**Comment by MSeifert04** [comment]
Well, I haven't checked if a change would be easy (generally changes in `io.fits` never are!) but it's probably doable. However I'm a bit concerned that **not** reordering them might produce FITS files that can't be handled by other FITS software. I haven't checked but FITS software can be pretty old. 😄 

So when reading & writing such a (col-)fits-plus file with `astropy.io.fits` the output-file isn't valid (col-)fits-plus anymore?

**Comment by olebole** [comment]
> So when reading & writing such a (col-)fits-plus file with astropy.io.fits the output-file isn't valid (col-)fits-plus anymore?

It looks like so. I maybe can convince @mbtaylor to change this for upcoming releases of STILTS/TOPCAT, even but then the same problem will be that the current installations will not be able to handle those files.

>  However I'm a bit concerned that not reordering them might produce FITS files that can't be handled by other FITS software

This is a dilemma, sure. You could introduce a flag somehwere indicating the standards version, so if one really needs an old standard, he can specify it.
However, usually the FITS standard experts are quite conservative, so hopefully they knew what they do when they relaxed this requirement. And the (col)fitsplus extension is already there for quite some time, without many interoperational problems (correct, @mbtaylor?).

**Comment by saimn** [comment]
The change should be quite easy as there is a specific method to handle this: https://github.com/astropy/astropy/blob/master/astropy/io/fits/hdu/hdulist.py#L787

But this raises a number of questions:

- Why not simply removing `EXTEND` as it is no more required after all ?
- If not removing `EXTEND`, where should it be placed ? There is no strict requirement here, but for practical reasons it's better if standard FITS keywords are grouped at the beginning of a header I think, and doing this without `NAXIS` can be more difficult.
- Changing this may break other softwares.
- Why changing every software to respect a non-standard convention from Topcat, instead of modifying Topcat directly ? Topcat could check the `VOTMETA` keyword presence without assumption on the order.
- How cfitsio handles this ?

**Comment by MSeifert04** [comment]
> The change should be quite easy

It almost never is in `io.fits`. You know 😄 

However I agree, the question if and how it should be changed is more complicated. Even though it isn't required anymore including `EXTEND` directly after the `NAXISn` keywords isn't forbidden (and for backwards-compatibility it is sort-of required).

**Comment by olebole** [comment]
> Why changing every software to respect a non-standard convention from Topcat

I agree that the implementation is a bit unfortunate here; however expecting a specific order of the keywords (including `EXTEND`) is fully conform to todays standard (3.0).

> Even though it isn't required anymore including EXTEND directly after the NAXISn keywords isn't forbidden

It is forbidden for (col)fits-plus files, as they are implemented today.

**Comment by mbtaylor** [comment]
Round-tripping [col]fits-plus FITS files through astropy, or other non-STIL-based software in general, is not a use case I would spend much (any?) effort on.

The FITS-plus convention (documented briefly e.g. at http://www.starlink.ac.uk/stil/sun252/outFitsplus.html) is in practice private to the [STIL](http://www.starlink.ac.uk/stil/) library.  The metadata relating to this convention is only usable by FITS-plus-aware software, which as far as I know means just things based on STIL, i.e. TOPCAT and STILTS.  Other software could in principle be modified to be FITS-plus-aware, but as far as I know has not been to date. This additional markup is in fact _intended_ to be invisible to non-FITS-plus-aware software, which by design should just see a normal, unadorned, FITS table.  The additional metadata may well get broken if the file/data is read, modified and written by some other software, so it's actually a good thing if a read-modify-write cycle by non-FITS-plus-aware software breaks the FITS-plus magic number.

So I do not support any change to astropy along these lines.

**Comment by olebole** [comment]
Is a workflow that uses astropy together with STILTS (or topcat) so unprobable?

**Comment by mbtaylor** [comment]
Not at all, and it will work without problems, but if you write a FITS-plus file from topcat and then read it into astropy, astropy will only see the basic FITS metadata, not the enhanced FITS-plus metadata.  That's bound to be the case unless astropy adds FITS-plus functionality (I've got no objection if they want to look into it, but I'm certainly not asking them to do that).  Given that astropy doesn't understand the additional metadata, it can't be expected to keep track of it, so it shouldn't be expected to to have to write it either.

**Comment by olebole** [comment]
The Fits-Plus extension is not hidden in astropy: it is an 1d array in the primary HDU, which is correctly identified in the header, and this will be read and written back by astropy (if not, that would be a bug IMO). And I would expect, that I can f.e. open a Fits-Plus file with astropy, add some headers, write it back and still have a fully compatible file -- this is at least what I would expect from the documentation of Fits-Plus.
A different question is the interpretation of the Fits-Plus array, but I don't see why it would be useless for astropy (or an extension of it) to support these metadata directly. This is however not the scope of this issue.
_(Sorry, that went to the wrong issue first)_

**Comment by MSeifert04** [comment]
@olebole Is it possible to share a FITS-plus file?

**Comment by mbtaylor** [comment]
A file following the FITS-plus convention is designed to work well if it is written by FITS-plus-aware software and read by FITS-plus-aware software.  So one user can write a FITS-plus file from TOPCAT, and another can read it using STILTS.   The second user benefits from the additional rich metadata that may have been added to it while it was in TOPCAT.  If a third user reads it using astropy, they can see all the data and basic metadata, but not rich metadata.

**Comment by olebole** [comment]
@MSeifert04 You may get one (11MB) here:
http://cdn.gea.esac.esa.int/Gaia/tgas_source/fits/TgasSource_000-000-015.fits

**Comment by olebole** [comment]
@mbtaylor sorry I don't see the big problem here. Fits-Plus isn't that magic, and neither the documentation nor the implementation would suggest that one can't add f.e. a header to the primary extension without breaking it. At least, if the order of the header keywords is kept.
And, ofcourse you can see and use the Fits-Plus metadata in astropy. Just straightforward created from your documentation:
```
import astropy.io
import io
vot = astropy.io.fits.open('TgasSource_000-000-015.fits')
metadata = astropy.io.votable.parse(io.BytesIO(vot[0].data))
print(metadata.get_first_table().description)
```

**Comment by MSeifert04** [comment]
So basically this _could_ (not saying that it necessarily **should**) be solved by keeping the position of th e`VOTMETA` keyword if it exists?

**Comment by olebole** [comment]
And the `COLFITS` keyword (which is, if present, [before the `VOTMETA` keyword](https://github.com/Starlink/starjava/blob/e27e8f0f86d313cde9ac110af62b0a43a15bf873/votable/src/main/uk/ac/starlink/votable/ColFitsPlusTableWriter.java#L47-L50). I am however not sure whether that specific case should be hardcoded, since next time someone else comes with another specific (uncommon, but allowed) keyword order. IMO the fits header should avoid specific cases and just stick to the standard.

**Comment by mbtaylor** [comment]
Reading it like that is perfectly OK.  But if you want to read it, modify it, and write it out again, you better understand FITS-plus well enough to know you're not messing things up.  Since there is not currently (AFAIK) code in astropy that can do that, the fact that you can't currently write out a slightly-modified FITS header in the way that FITS-plus likes is not a current problem in astropy as far as FITS-plus goes (though it's possible it might present some problems in future given that it's enforcing something no longer required by the FITS standard).

The code in STIL that writes FITS-plus generates the output FITS file from scratch based on its internal (format-neutral) table data model.  It does not attempt to remember the FITS headers it saw if/when it read the file on input.  I.e. even STIL doesn't attempt the difficult trick of read/modify/write a file in this format at the FITS level.
So I think it is respectable (even beneficial) to impose my own magic number for this convention rather than conform to other formats that look a bit like it - I only want to pick up things that are explicitly asserting that they are conforming to this convention.

**Comment by olebole** [comment]
Apart from the missing in-depth documentation for Fits-Plus, there is no reason why one should not be able to modify or create such a file with the help of astropy. The (col)fits-plus code is quite unchanged over the years in the starjava repository, and you obviously care much about incompatible changes. So if someone has a real use case here (or if a volunteer just wants to donate code to `astropy.io.votable` ;-) ), (s)he should be able to dig into the code to extract the specification and then just start implementing it. Astropy should not prevent from doing so.

**Comment by mbtaylor** [comment]
Yes, I agree with that, it's a potential problem if astropy or an astropy user wants to introduce FITS-plus-awareness.  I'm just saying that I don't think it's a current problem in practice with (non-FITS-plus-aware) astropy.

**Comment by olebole** [comment]
But isn't this a small step only? As far as I can see, it is (for reading FitsPlus) just taking the metadata from the primary header, create a VO table(s) out if, and then adding the content of the extensions, right? And for writing just putting the UTF-8 encoded byte array of a VOTable containing only the metadata into the primary header and adding the `VOTMETA` (and maybe `COLFITS`) keywords? Do I miss something important here?

**Comment by mbtaylor** [comment]
Yes that's about right.  If you want to implement I/O handlers for this format in astropy or any other software, go ahead.

There is an action on me to write up the details of this format for reference in the [registry of FITS conventions](https://fits.gsfc.nasa.gov/fits_conventions.html).  If anybody is keen to implement FITS-plus outside of STIL I will bump that up my list of priorities.

**Comment by olebole** [comment]
@MSeifert04 On the [Java side](https://github.com/nom-tam-fits/nom-tam-fits), there was a [comment](https://github.com/nom-tam-fits/nom-tam-fits/issues/113#issuecomment-287767045) by @TomMcGlynn that may apply here as well:

> The header ordering was put in place to address a FITS mandate that has since been revoked. While it was perfectly natural to continue to sort the EXTEND header cards that way to accommodate users who might still be assuming the old requirement, it's been a really long time and it's now causing a certain degree of grief. Changing HeaderOrder to leave the EXTEND keyword alone, but enforce the ordering requirements that FITS still has seems simple and appropriate. Users who want to old behavior can either ensure that the EXTEND keyword is written immedately after NAXISn or use an old version of the library. If users really want to write invalid FITS files they can always write there own version of HeaderOrder.

**Comment by MSeifert04** [comment]
I'm actually +/- 0 on a change here. I can see the benefit of not modifying the header "by magic", however I'm wary that we break compatibility with other software (there's a huge amount of scripts, legacy software around that isn't "maintained"). 

@pllim Do you have a preference?

**Comment by pllim** [comment]
I would prefer not to change it. When we have a proper maintainer who knows the FITS standards very well, then we can revisit.

**Comment by olebole** [comment]
From the standard, it is IMO quite clear: there is no rule anymore about the order of the `EXTEND` keyword. The problem is that topcat/stilts is quite popular, and as long as this is unfixed, astropy will have no chance to write their "Fits-Plus" format although it is standard conform FITS. I am just a bit worried about interoperability with a commonly used tool here.

**Comment by pllim** [comment]
@olebole , unfortunately "there is no rule anymore" does not guarantee full backward-compatibility, as pointed out above. It is not a can of worms I want to deal with as "deputy" right now...

**Comment by olebole** [comment]
What about an optional `standard-version` keyword that allows to specify the standard? This should make everyone happy.

**Comment by pllim** [comment]
> What about an optional `standard-version` parameter

That might be a plausible compromise, as long as default does not change current behavior. A PR is welcome.

**Comment by olebole** [comment]
After thinking a bit about this, I am not sure anymore: adding such a parameter would imply that Astropy then writes according to a specific standard, which goes far beyond the order of header keywords. Also, if we take an older standard version as default, then some current features of astropy would not be supported because they require a newer version (I didn't check that yet, however). So, adding `standards_version` would just open another can of worms. We could add a `sort_header` parameter instead, but that may be a bit too special (what do you think?)
Another point: The re-ordering of keywords before writing them is usually not required at all: The `PrimaryHDU` class already creates them in the correct order, and if it was read from a file, then it would be wise to not touch the header at all. So, unless the user explicitly changes the order, the header remains compliant to version < 3. And when explicitly changing the header order, one is responsible for that anyway.
So, my first proposal would be still just to drop the explicit ordering before writing and to rely on the order that was correctly created initially.

**Comment by pllim** [comment]
> So, my first proposal would be still just to drop the explicit ordering before writing and to rely on the order that was correctly created initially.

There is a big assumption here that the order "was correctly created initially." In your case, yes. But we also have to deal with a bunch of legacy files, where that might not be the case.

Also, "sort_header" sounds too general (it sounds like you might use it to sort alphabetically or something). I don't have a good alternative name to give you, unfortunately.

**Comment by olebole** [comment]
There is no "correct order" anymore by standard, at least with respect to `EXTEND`. Files without `EXTEND` right after `NAXIS*` are not valid according to FITSv2, but they valid according FITsv3. So, if we would not re-order `EXTEND`, we would not change the validity of the file. Which is IMO correct, since the order may matter and in the case of FitsPlus it *does* matter: Although FitsPlus is a absolutely valid format, a simple sequence by reading and writing the file makes it unusable:
```
from astropy.io import fits
f = fits.open('in.fits')
f.writeto('out.fits')
```
IMO this should **never** make a valid file unusable. 
If someone really needs the header in a specific order, he still can sort explicitly.
At least, I would not suspect from astropy to destroy correct files in order to automatically "repair" other files with respect to a standard that is outdated since 7 years.
Again: the only use case that is currently there is to "repair" files. Files conform to the old standard will be written according to this standard, and also newly created files conform to the old standard unless the order is explicitly changed. But files that are conform to the new standard should be processed flawlessly and without the risk to destroy them. Do you really have a different opinion here?

**Comment by pllim** [comment]
> Do you really have a different opinion here?

If you put it that way, not really. But I am just a "deputy". Perhaps ask the Astropy user mailing list in case there is any strong objection?

**Comment by olebole** [comment]
Maybe @embray could have a look here? The `EXTEND` handling dates back 5 years from now in [pyfits](https://github.com/spacetelescope/PyFITS/commit/453befbba0ab0dd8bded50ad744fd9b7b1609f4b).

**Comment by embray** [comment]
> It looks like so. I maybe can convince @mbtaylor to change this for upcoming releases of STILTS/TOPCAT, even but then the same problem will be that the current installations will not be able to handle those files.

I think these special formats should be changed to be compatible with older FITS versions that made some specification about this.  Because "once FITS always FITS", right? ;) ;) ;)

**Comment by embray** [comment]
> What about an optional standard-version parameter that allows to specify the standard? This should make everyone happy.

I tried to do that at one point.  It was going to be a feature of my FITS validation framework, to allow validation according to different versions of the FITS standard.  Unfortunately, in practice, it seems understanding of "the FITS standard" and what that is is very unevenly applied and often nonsensical.  

What really needs to be done here is overhauling of the currently rather broken and convoluted verification framework in pyfits (long on my TODO list but rarely a priority) to make it easier, when writing a file, to prevent changes to the file.  To some extent you can do this by specifying the appropriate `output_verify` option (e.g. `output_verify='ignore'`) but it isn't always obeyed properly.  Have you given that a try for this case?

**Comment by mbtaylor** [comment]
I'm still thinking about whether the greater good would be served by changing or not changing this in future versions of STIL. If I was writing it for the first time now, I don't think I would require a particular order, since it seems a bit contrary to the spirit of normal FITS header handling.  But in the mean time: the code that's currently in place does not violate "once FITS always FITS".

**Comment by olebole** [comment]
@embray 
> I think these special formats should be changed to be compatible with older FITS versions that made some specification about this. Because "once FITS always FITS", right? ;) ;) ;)

IMO this slogan means that any file that was a FITS file in the past shall stay a valid FITS file forever, not that any old FITS reader will always be able to read all modern FITS files. So, there are valid FITS files (for example the FitsPlus format from TOPCAT) which are not readable by old readers -- but these files are still standard conform (and should be supported by astropy).

>To some extent you can do this by specifying the appropriate output_verify option (e.g. `output_verify='ignore'`) but it isn't always obeyed properly. Have you given that a try for this case?

Doesn't help:
```
>>> from astropy.io import fits
>>> p = fits.PrimaryHDU()
>>> p.header.insert(3, ('VOTMETA', True))
>>> list(p.header)
['SIMPLE', 'BITPIX', 'NAXIS', 'VOTMETA', 'EXTEND']
>>> p.writeto('t.fits', output_verify='ignore')
>>> list(p.header)
['SIMPLE', 'BITPIX', 'NAXIS', 'EXTEND', 'VOTMETA']
```

**Comment by embray** [comment]
> IMO this slogan means that any file that was a FITS file in the past shall stay a valid FITS file forever, not that any old FITS reader will always be able to read all modern FITS files.

I think that's certainly true, though I also see it misapplied frequently to mean the opposite, or in ways that stymie improving the format.

```
>>> from astropy.io import fits
>>> p = fits.PrimaryHDU()
>>> p.header.insert(3, ('VOTMETA', True))
>>> list(p.header)
['SIMPLE', 'BITPIX', 'NAXIS', 'VOTMETA', 'EXTEND']
>>> p.writeto('t.fits', output_verify='ignore')
>>> list(p.header)
['SIMPLE', 'BITPIX', 'NAXIS', 'EXTEND', 'VOTMETA']
```

Okay, thanks for the example.  I do think this is sort of a bug but I'm not sure.  It looks like that for Image HDUs the 'EXTEND' card is being handled in two separate places in a sort of duplicate manner.  First here:

https://github.com/astropy/astropy/blob/a2d0fe7944153a4c87785537ea8416918b52c6d8/astropy/io/fits/hdu/image.py#L1013

then here:

https://github.com/astropy/astropy/blob/a2d0fe7944153a4c87785537ea8416918b52c6d8/astropy/io/fits/hdu/image.py#L1024

The latter is where output verification would be handled.  So normally, if somehow the `EXTEND` keyword were out of place, this would in principle correct it.  However, it also ''always'' calls `update_header` before writing, which moves the `EXTEND` keyword to the "normally" correct location.

I think it might suffice to just delete that `update_header` method entirely, but I haven't tested yet.  I'd be fine with that since the default behavior would still be to fix its location, but we would allow it to be left unfixed as well.  This was just one example where the code was being very aggressive about enforcing the standard, but perhaps in a way that isn't terribly important?

**Comment by olebole** [comment]
I tried t remove the `update_header` method; however then the `TestImageFunctions.test_rewriting_large_scaled_image` stumbles upon writing a modified file originating from `astropy/io/fits/tests/data/fixed-1890.fits`:
```
E               VerifyError: 
E               Verification reported errors:
E               HDU 0:
E                   'EXTEND' card at the wrong place (card 7).
```
This file has indeed the `EXTEND` keyword on a different place:
```
SIMPLE  =                    T          / Standard FITS (NOST-100-2.0)
BITPIX  =                   16 / # of bits per pix value                        
NAXIS   =                    2 / # of axes in data array                        
NAXIS1  =                  100 / # of pixels in <axis direction>                
NAXIS2  =                  100 / # of pixels in <axis direction>                
BSCALE  =                    1                                                  
BZERO   =                32768                                                  
EXTEND  =                    T          / FITS Extension may be present         
```
Interestingly, the file claims to be [`NOST-100-2.0`](http://www.aanda.org/articles/aa/full/2001/34/aah2901/aah2901.html), which is wrong because of the placing for the `EXTEND` keyword.
So, either the `update_header()` method must obey the `verify_output` flat (... and the flag name is a bit misleading then, since it does not just verify the output, but also may change it), or the header order check must be changed to not test the place of `EXTEND` (what I would prefer, since `EXTEND` is not required to be at a specific position anymore).

**Comment by embray** [comment]
This all speaks to why the "verification"/"fix" framework needs overhaul.  There probably was a good reason for this at one time, but it should be easier to enable/disable specific checks (sort of like with a linter).

I think there may be a middle ground?  The test that fails is certainly assuming that `update_header` method is implemented, and is hence always moving the `EXTEND` keyword in the header (I guess the test file had it in the wrong position in the first place, as you showed).

This could instead be changed to a warning, not an outright error, and left unfixed by default, but still fix it if `output_verify='fix'` is set.
